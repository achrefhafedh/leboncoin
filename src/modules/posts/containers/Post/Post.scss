@import '../../../../assets/colors.scss';

.post {
  padding: 20px;
  .success {
    display: block;
    text-align: center;
    padding: 10px;
    color: $white;
    background: $blue;
    margin: 20px 0px;
  }
  textarea {
    width: 100%;
    height: 200px;
    display: block;
    margin-bottom: 20px;
    font-size: 1.2rem;
  }
  .action {
    display: flex;
    .containerCheckbox {
      flex: 1;
      font-size: 1rem;
      cursor: pointer;
      text-align: right;
      label,
      input {
        cursor: pointer;
      }
    }
    .btn {
      display: block;
      background: $blue;
      border: 1px solid darken($blue, 5%);
      padding: 1em 2em;
      color: $white;
      margin-right: 0.5em;
      font-size: 0.8rem;
      box-shadow: inset 0px 1px 0px fadeout($white, 80%);
      cursor: pointer;
      &:hover {
        background: lighten($blue, 5%);
      }
      &:active {
        background: $blue;
        box-shadow: inset 0px 1px 1px fadeout($black, 90%);
      }
      &:focus {
        outline: none;
      }
    }
  }
}
