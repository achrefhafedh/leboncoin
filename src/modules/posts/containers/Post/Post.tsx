import React, { Component, FormEvent } from 'react';
import { connect } from 'react-redux';
import { ThunkDispatch } from 'redux-thunk';
import { getModuleState } from 'redux-register-module';

import { Store } from 'store';
import { addPost } from '../../redux/actions';

import { postsHeaderLayoutHOC } from '../../hoc/postsHeaderLayout';
import { PostType } from '../../types';

import './Post.scss';

type OwnProps = {};

type StateProps = {
  userType: string;
};

type DispatchProps = {
  post: (post: PostType) => void;
};

type PostProps = StateProps & DispatchProps & OwnProps;

type PostState = {
  checkboxDisabled: boolean;
  message: string;
  publicStatus: boolean;
  validPost: boolean;
};

export class PostContainer extends Component<PostProps, PostState> {
  constructor(props: PostProps) {
    super(props);
    this.state = {
      checkboxDisabled: false,
      message: '',
      publicStatus: true,
      validPost: false,
    };
  }

  componentDidMount() {
    const { userType } = this.props;
    if (userType && userType === 'guest') {
      this.setState(() => ({
        checkboxDisabled: true,
      }));
    }
  }

  postHandler = () => {
    const { message, publicStatus } = this.state;
    const { post } = this.props;
    const params = {
      content: message,
      type: publicStatus ? 'public' : 'private',
    };
    post(params);
    this.setState(() => ({
      message: '',
      validPost: true,
    }));
  };

  messageHandler = (event: FormEvent<HTMLTextAreaElement>) => {
    const message = event.currentTarget.value;
    this.setState(() => ({
      message,
    }));
  };

  statusHandler = () => {
    const { publicStatus } = this.state;
    this.setState(() => ({
      publicStatus: !publicStatus,
    }));
  };

  renderContent = () => (
    <div className="post">
      <h2> Add Your Post :</h2>
      <textarea
        value={this.state.message}
        onChange={event => this.messageHandler(event)}
      />
      {/* TODO: for Button Label and Bheckbox we need an external module contain all component HTML */}
      <div className="action">
        <button onClick={this.postHandler} className="btn">
          Post
        </button>
        {!this.state.checkboxDisabled && (
          <div className="containerCheckbox">
            <input
              onChange={this.statusHandler}
              type="checkbox"
              id="private"
              name="private"
              value="private"
            />
            <label htmlFor="private">Private</label>
          </div>
        )}
      </div>
      {this.state.validPost && <div className="success">Success Post</div>}
    </div>
  );
  render() {
    return this.renderContent();
  }
}

const mapStateToProps = (state: Store, onProps: OwnProps): StateProps => {
  return {
    userType: getModuleState('login', state).user.type,
  };
};

const mapDispatchToProps = (
  dispatch: ThunkDispatch<Store, any, any>
): DispatchProps => ({
  post: (post: PostType) => {
    dispatch(addPost(post));
  },
});

const PostConnected = connect(
  mapStateToProps,
  mapDispatchToProps
)(PostContainer);

export const Post = postsHeaderLayoutHOC(<PostConnected />);
