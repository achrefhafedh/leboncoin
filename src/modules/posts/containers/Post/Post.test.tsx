import React from 'react';
import { configure, shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

import { PostContainer } from './Post';

configure({ adapter: new Adapter() });

describe('render without crashing', () => {
  let userType: string;
  let post: () => void;
  beforeEach(() => {
    userType = 'guest';
    post = jest.fn();
  });

  it('render without crashing', () => {
    const postItem = shallow(<PostContainer post={post} userType={userType} />);
    expect(postItem).toMatchSnapshot();
  });

  it('execute function postHandler', () => {
    const wrapper = shallow<PostContainer>(
      <PostContainer post={post} userType={userType} />
    );
    wrapper.instance().postHandler();
  });

  it('execute function messageHandler', () => {
    const event = {
      currentTarget: {
        value: 'value',
      },
    } as React.ChangeEvent<HTMLTextAreaElement>;
    const wrapper = shallow<PostContainer>(
      <PostContainer post={post} userType={userType} />
    );
    wrapper.instance().messageHandler(event);
  });

  it('execute function statusHandler', () => {
    const wrapper = shallow<PostContainer>(
      <PostContainer post={post} userType={userType} />
    );
    wrapper.instance().statusHandler();
  });
});
