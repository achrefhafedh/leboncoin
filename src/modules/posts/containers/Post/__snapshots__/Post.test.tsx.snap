// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render without crashing render without crashing 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <PostContainer
    post={[MockFunction]}
    userType="guest"
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <h2>
           Add Your Post :
        </h2>,
        <textarea
          onChange={[Function]}
          value=""
        />,
        <div
          className="action"
        >
          <button
            className="btn"
            onClick={[Function]}
          >
            Post
          </button>
        </div>,
        false,
      ],
      "className": "post",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": " Add Your Post :",
        },
        "ref": null,
        "rendered": " Add Your Post :",
        "type": "h2",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "onChange": [Function],
          "value": "",
        },
        "ref": null,
        "rendered": null,
        "type": "textarea",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <button
              className="btn"
              onClick={[Function]}
            >
              Post
            </button>,
            false,
          ],
          "className": "action",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Post",
              "className": "btn",
              "onClick": [Function],
            },
            "ref": null,
            "rendered": "Post",
            "type": "button",
          },
          false,
        ],
        "type": "div",
      },
      false,
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <h2>
             Add Your Post :
          </h2>,
          <textarea
            onChange={[Function]}
            value=""
          />,
          <div
            className="action"
          >
            <button
              className="btn"
              onClick={[Function]}
            >
              Post
            </button>
          </div>,
          false,
        ],
        "className": "post",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": " Add Your Post :",
          },
          "ref": null,
          "rendered": " Add Your Post :",
          "type": "h2",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "onChange": [Function],
            "value": "",
          },
          "ref": null,
          "rendered": null,
          "type": "textarea",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <button
                className="btn"
                onClick={[Function]}
              >
                Post
              </button>,
              false,
            ],
            "className": "action",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Post",
                "className": "btn",
                "onClick": [Function],
              },
              "ref": null,
              "rendered": "Post",
              "type": "button",
            },
            false,
          ],
          "type": "div",
        },
        false,
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
